server/src/main/java/org/opensearch/gateway/ShardCache.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway;
+
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.logging.Loggers;
+import org.opensearch.core.index.shard.ShardId;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import reactor.util.annotation.NonNull;
+
+/**
+ * Cache implementation of transport actions returning single shard related data in the response.
+ *
+ * @param <K> Response type of transport action.
+ */
+public class ShardCache<K extends BaseNodeResponse> extends BaseShardCache<K> {
```

Comment:
IS this cache expected to be thread safe?

Reply:
Hi @Bukhtawar,

No. Currently AsyncShardFetch class uses a normal HashMap and this flow is executed via a pending task which is a single threaded execution. Also put in the map only happens at the time of init and when responses are returned by data nodes. So, 2 threads will never try to write into this map in parallel.

Re-iterating - this PR just does the code abstraction. No functionalities are changed.